dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,346 @@ module Dhall.LSP.Handlers where
-initializedHandler :: LSP.LspFuncs () -> J.InitializedNotification -> IO ()
+-- Workaround to make our single-threaded LSP fit dhall-lsp's API, which
+-- expects a multi-threaded implementation. Reports errors to the user via the
+-- LSP `ShowMessage` notification.
+wrapHandler
+  :: MVar ServerState
+  -> (a -> HandlerM ())
+  -> a
+  -> IO ()
+wrapHandler vstate handle message =
+  modifyMVar_ vstate $
+    execStateT . runExceptT $
+      catchE (handle message) lspUserMessage
+
+lspUserMessage :: (Severity, Text) -> HandlerM ()
+lspUserMessage (Log, text) =
+  lspSendNotification LSP.NotLogMessage J.WindowLogMessage
+    $ J.LogMessageParams J.MtLog text
+lspUserMessage (severity, text) =
+  lspSendNotification LSP.NotShowMessage J.WindowShowMessage
+    $ J.ShowMessageParams severity' text
+  where severity' = case severity of
+          Error -> J.MtError
+          Warning -> J.MtWarning
+          Info -> J.MtInfo
+          Log -> J.MtLog
+
+
+lspSend :: LSP.FromServerMessage -> HandlerM ()
+lspSend msg = do
+  send <- lift $ gets (view (lspFuncs . sendFunc))
+  liftIO $ send msg
+
+lspRespond :: (J.ResponseMessage response -> LSP.FromServerMessage)
+  -> J.RequestMessage J.ClientMethod request response -> response -> HandlerM ()
+lspRespond constructor request response =
+  lspSend . constructor $ LSP.makeResponseMessage request response
+
+lspSendNotification
+  :: (J.NotificationMessage J.ServerMethod params -> LSP.FromServerMessage)
+  -> J.ServerMethod -> params -> HandlerM ()
+lspSendNotification constructor method params =
+  lspSend . constructor $ J.NotificationMessage "2.0" method params
+
+lspRequest
+  :: (J.RequestMessage J.ServerMethod params response -> LSP.FromServerMessage)
+  -> J.ServerMethod -> params -> HandlerM ()
+lspRequest constructor method params = do
+  getNextReqId <- lift $ gets (LSP.getNextReqId . view lspFuncs)
+  reqId <- liftIO getNextReqId
+  lspSend . constructor $ J.RequestMessage "2.0" reqId method params
+
+-- | A helper function to query haskell-lsp's VFS.
+readUri :: J.Uri -> HandlerM Text
+readUri uri = do
+  getVirtualFileFunc <- lift $ gets (LSP.getVirtualFileFunc . view lspFuncs)
+  asd <- liftIO $ getVirtualFileFunc uri
+  case asd of
+    Just (LSP.VirtualFile _ rope) -> return (Rope.toText rope)
+    Nothing -> fail $ "Could not find " <> show uri <> " in VFS."
+
+loadFile :: J.Uri -> HandlerM (Expr Src X)
+loadFile uri = do
+  txt <- readUri uri
+  fileIdentifier <- fileIdentifierFromUri uri
+  cache <- lift $ gets (view importCache)
+
+  expr <- case parse txt of
+    Right e -> return e
+    _ -> throwE (Error, "Failed to parse Dhall file.")
+
+  loaded <- liftIO $ load fileIdentifier expr cache
+  (cache', expr') <- case loaded of
+    Right x -> return x
+    _ -> throwE (Error, "Failed to resolve imports.")
+  -- Update cache. Don't cache current expression because it might not have been
+  -- written to disk yet (readUri reads from the VFS).
+  lift $ modify (set importCache cache')
```

Comment:
Similarly, here you can use `assign` or `(.=`):

```haskell
assign importCache cache'
```

```haskell
importCache .= cache'
```

See: http://hackage.haskell.org/package/lens-family-1.2.3/docs/Lens-Family2-State-Lazy.html#v:assign